File: 96.不同的二叉搜索树.cpp
/*
 * @lc app=leetcode.cn id=96 lang=cpp
 *
 * [96] 不同的二叉搜索树
 */
#include <vector>
using namespace std;
// @lc code=start
class Solution
{
public:
    int numTrees(int n)
    {
        vector<int> dp(n + 1, 0);
        dp[0] = 1;
        // 相当于0个节点也是一种

        for (int i = 1; i <= n; i++)
        {
            for (int j = 1; j <= i; j++)
            {
                dp[i] += dp[j - 1] * dp[i - j];
                // 左边j-1，右边i-j
                // 相乘是因为不同的组合就是不同的树
            }
        }
        return dp[n];
    }
};
// @lc code=end

File: 96.不同的二叉搜索树.py
#
# @lc app=leetcode.cn id=96 lang=python3
#
# [96] 不同的二叉搜索树
#


# @lc code=start
class Solution:
    def numTrees(self, n: int) -> int:
        dp = [0] * (n + 1)
        dp[0] = 1

        for i in range(1, n + 1):
            # 枚举root节点
            for root in range(1, i + 1):
                left_count = root - 1
                right_count = i - root

                dp[i] += dp[left_count] * dp[right_count]
        
        return dp[n]


# @lc code=end
sol = Solution()
res = sol.numTrees(n = 3)
print(res)